bazel: 2.0.0 -> 2.1.0

jpgneves opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

The official Bazel wrapper script included from version 2.0.0 onwards enforces the actual `bazel` binary (previously named `bazel-real` in this package) has the following naming convention, if the root of your Bazel project contains a `.bazelversion` file: `bazel-${version}-${os}-${arch}`.

This modifies output binary name and, while we're at it, upgrades Bazel to 2.1.0.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jpgneves committed:
bazel: 2.0.0 -> 2.1.0

uri-canva commented:
Looking at the current implementation of the script on master, it doesn't enforce that naming convention for the binary, it still supports calling it `bazel-real` as well: https://github.com/bazelbuild/bazel/blob/59d7864b3a39008e6b4d1447abcdc59cd9906e88/scripts/packages/bazel.sh

In that case the `.bazelversion` becomes a check that makes sure the version you're running matches it.

Profpatsch commented:
> In that case the `.bazelversion` becomes a check that makes sure the version you're running matches it.

That look what we should be doing.

andir commented:
Please don't just update this version. Does 2.0 remain compatible with 2.1 in all cases? If not we should consider having another attribute within nixpkgs. The update to Bazel 2.0 already caused some breakage that we didn't get rid of yet. See the relevant conversations in https://github.com/NixOS/nixpkgs/pull/76851 and my WIP PR #80953.

Profpatsch commented:
bazel [promises no backwards-compatible changes for three months at a time](https://blog.bazel.build/2020/02/06/bazel-2.1.html)

If they release a breaking change, they bump the major version.

Profpatsch commented:
> This modifies output binary name and, while we're at it, upgrades Bazel to 2.1.0.

Let’s put this in another PR

jpgneves commented:
Will do! Apologies for the messy intermixing of changes. Should I close this PR and resubmit separate ones, or is it sufficient to create a new one with the version change?

Profpatsch commented:
> Should I close this PR and resubmit separate ones, or is it sufficient to create a new one with the version change?

It’s fine if you pull it out in a separate commit, they can both be in this PR.

Profpatsch commented:
related comment in the original issue:

https://github.com/NixOS/nixpkgs/issues/80950#issuecomment-590337389

Profpatsch commented:
So I guess as long as we have a normal `bazel` binary and the versioning is what bazel expects, it’s better to be as specific as possible.

Profpatsch reviewed (approved):
Once the two commits are split, this is good to go. :+1:

jpgneves committed:
bazel: output correct binary name

From Bazel 2.0.0 onwards, Bazel looks for a binary named
`bazel-${version}-${os_arch}` if the project root contains a
`.bazelversion` file or the USE_BAZEL_VERSION environment
variable is set.

This change ensures we output a binary with the correct name
for the current version and OS/arch combination.

Profpatsch commented:
@GrahamcOfBorg build bazel.tests

jpgneves commented:
Regarding the failed tests on darwin, I do not have a machine where I can work to fix the issue, so I'd appreciate any assistance in figuring it out, if it's a blocker (which it sounds like it is). :)

Profpatsch commented:
@kalbasit you seem to have fixed the protobuf test last time, could you take a look?

uri-canva commented:
The failure I'm getting on darwin is the one that was fixed in https://github.com/NixOS/nixpkgs/pull/80635.

uri-canva commented:
Rebased on current master (b1ec189c9faa017e7a8dc3aa3b8243da3d61b8c7) it works.

jpgneves commented:
Rebased it on current master right now.

Profpatsch commented:
cool
